### ISSUE TITLE
buildah run $container not running command after specifying command using buildah config --cmd

### ISSUE THREAD

**Issue by rfairley** [projectatomic/buildah#733]
**Description**

While following the tutorial https://www.projectatomic.io/blog/2017/11/getting-started-with-buildah/, at one point it asks to specify a command using `buildah config`, which should allow this command to be executed without specifying later in `buildah run`. `buildah run` asks for the command to be specified, instead of executing the command that was specified earlier. With `docker run` on the same image, the command is executed without specifying.

**Steps to reproduce the issue:**

1. Build a container from scratch, mount, and install dependencies onto it:

```
# newcontainer=$(buildah from scratch)
# scratchmnt=$(buildah mount $newcontainer)
# dnf install --installroot $scratchmnt --release 26 bash coreutils --setopt install_weak_deps=false -y
```

2. Create a script, and configure this to run as a command upon running the container with `buildah run`.
```
# touch runecho.sh
# vi runecho.sh
```

paste the following into `runecho.sh`:

```
#!/bin/bash
for i in `seq 0 9`;
do
    echo "This is a new container from ipbabble [" $i "]"
done
```
```
# chmod +x runecho.sh
# buildah copy $newcontainer ./runecho.sh /usr/bin
# buildah config --cmd /usr/bin/runecho.sh $newcontainer
```

3. Run the container with `buildah run`. This is where the bug occurs - the script `runecho.sh` should run at this point since the command was specified earlier.

```
# buildah run $newcontainer
```

Specifying the `runecho.sh` script does run it:

```
# buildah run $newcontainer /usr/bin/runecho.sh
```

4. Committing the container to an image, pushing to the docker daemon, and running with `docker run` runs `runecho.sh` as expected.

```
# buildah unmount $newcontainer
# buildah commit $newcontainer fedora-bashecho
# dnf -y install docker
# systemctl start docker
# buildah push fedora-bashecho docker-daemon:fedora-bashecho:latest
# docker run fedora-bashecho

```
**Describe the results you received:**

`buildah run $newcontainer` gives output:

```
command must be specified
```

`buildah run $newcontainer /usr/bin/runecho.sh` and `docker run fedora-bashecho` both give output:

```
This is a new container from ipbabble [ 0 ]
This is a new container from ipbabble [ 1 ]
This is a new container from ipbabble [ 2 ]
This is a new container from ipbabble [ 3 ]
This is a new container from ipbabble [ 4 ]
This is a new container from ipbabble [ 5 ]
This is a new container from ipbabble [ 6 ]
This is a new container from ipbabble [ 7 ]
This is a new container from ipbabble [ 8 ]
This is a new container from ipbabble [ 9 ]
```

`buildah inspect $newcontainer` also shows that the command appears in the config:
```
{
    "Type": "buildah 0.0.1",
    "FromImage": "",
    "FromImageID": "",
    "Config": "",
    "Manifest": "",
    "Container": "working-container",
    "ContainerID": "04fcb63390e52450bedff150c6c91c1cf3dc0ded81ff242acbccb12be82c92e5",
    "MountPoint": "",
    "ProcessLabel": "system_u:system_r:container_t:s0:c738,c879",
    "MountLabel": "",
    "ImageAnnotations": null,
    "ImageCreatedBy": "",
    "OCIv1": {
        "created": "2018-05-29T14:39:16.078110841Z",
        "architecture": "amd64",
        "os": "linux",
        "config": {
            "Cmd": [
                "/usr/bin/runecho.sh"
            ],
            "WorkingDir": "/"
        },
        "rootfs": {
            "type": "",
            "diff_ids": null
        }
    },
    "Docker": {
        "created": "2018-05-29T14:39:16.078110841Z",
        "container_config": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": null,
            "Cmd": [
                "/usr/bin/runecho.sh"
            ],
            "Image": "",
            "Volumes": null,
            "WorkingDir": "/",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": null
        },
        "config": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": null,
            "Cmd": [
                "/usr/bin/runecho.sh"
            ],
            "Image": "",
            "Volumes": null,
            "WorkingDir": "/",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": null
        },
        "architecture": "amd64",
        "os": "linux"
    },
    "DefaultMountsFilePath": ""
}
```

**Describe the results you expected:**

`buildah run $newcontainer`, `buildah run $newcontainer /usr/bin/runecho.sh`, and `docker run fedora-bashecho` should all give the following output:

```
This is a new container from ipbabble [ 0 ]
This is a new container from ipbabble [ 1 ]
This is a new container from ipbabble [ 2 ]
This is a new container from ipbabble [ 3 ]
This is a new container from ipbabble [ 4 ]
This is a new container from ipbabble [ 5 ]
This is a new container from ipbabble [ 6 ]
This is a new container from ipbabble [ 7 ]
This is a new container from ipbabble [ 8 ]
This is a new container from ipbabble [ 9 ]
```

**Output of `rpm -q buildah` or `apt list buildah`:**

```
buildah-1.0-1.git1ab80bc.fc28.x86_64
```

**Output of `buildah version`:**

```
Version:       1.0
Go Version:    go1.10.1
Image Spec:    1.0.0
Runtime Spec:  1.0.0
Git Commit:    1ab80bc
Built:         Mon May  7 14:01:02 2018
OS/Arch:       linux/amd64
```

**Output of `cat /etc/*release`:**

```
Fedora release 28 (Twenty Eight)
NAME=Fedora
VERSION="28 (Cloud Edition)"
ID=fedora
VERSION_ID=28
PLATFORM_ID="platform:f28"
PRETTY_NAME="Fedora 28 (Cloud Edition)"
ANSI_COLOR="0;34"
CPE_NAME="cpe:/o:fedoraproject:fedora:28"
HOME_URL="https://fedoraproject.org/"
SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=28
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=28
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="Cloud Edition"
VARIANT_ID=cloud
Fedora release 28 (Twenty Eight)
Fedora release 28 (Twenty Eight)
```

**Output of `uname -a`:**

```
Linux localhost.localdomain 4.16.3-301.fc28.x86_64 #1 SMP Mon Apr 23 21:59:58 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# storage.conf is the configuration file for all tools
# that share the containers/storage libraries
# See man 5 containers-storage.conf for more information

# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver
driver = "overlay"

# Temporary storage location
runroot = "/var/run/containers/storage"

# Primary Read/Write location of container storage
graphroot = "/var/lib/containers/storage"

[storage.options]
# AdditionalImageStores is used to pass paths to additional Read/Only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Size is used to set a maximum size of the container image.  Only supported by
# certain container storage drivers.
size = ""

# OverrideKernelCheck tells the driver to ignore kernel checks based on kernel version
override_kernel_check = "true"
```

**Comment by rfairley** [comment]
I see, thank you for clarifying that - now I am more clear on how buildah compares to the Dockerfile. I understand that buildah run does not perform the same function as docker run or podman run.

The main point I would like to address in the issue is that `buildah run` should run the command previously specified using `buildah config` - as the tutorial linked showed.

**Comment by TomSweeneyRedHat** [comment]
@rhatdan and @rfairley  The tutorial is out of spec now, @ipabble is planning to touch it up, but has been globe hopping as of late.  I'll see if I can touch it up in the short term and let @ipbabble fix from there.

**Comment by ipbabble** [comment]
Yes. Sorry this needs to be fixed. And a few other wee things too. Next
week.

William Henry
Senior Consulting Software Engineer
Portfolio Architecture
Red Hat
Colorado, USA.

**Comment by rfairley** [comment]
Sounds good! Thanks.

**Comment by TomSweeneyRedHat** [comment]
Closing this issue as I believe it's been addressed via the changes in #733 and I think @ipbabble will be doing some other touchups next week.  @rfairley if you think this problem still needs more help, please reopen this or start up a new issue.  Thanks again for pointing this out to us.

**Comment by rhatdan** [comment]
Thanks @rfairley for keeping us honest and pointing out where we have problems in our documentation.
